extensions/amp-analytics/0.1/requests.js

```diff
@@ -363,113 +313,93 @@ export class RequestHandler {
+    return urlReplacementService.expandStringAsync(base, bindings);
   });
   if (msg.indexOf('${extraUrlParams}') < 0) {
     // No need to append extraUrlParams
     return basePromise;
   }
 
-  //return base url with the appended extra url params;
-  const extraUrlParamsStrPromise = getExtraUrlParams(
-      variableService, configParams, trigger['extraUrlParams'], expansionOption)
-      .then(params => {
-        const str = getExtraUrlParamsString(params);
-        return urlReplacementService.expandUrlAsync(str, bindings);
-      });
-
   return basePromise.then(expandedMsg => {
-    return constructExtraUrlParamStrs(expandedMsg, [extraUrlParamsStrPromise]);
+    const params = Object.assign({}, configParams, trigger['extraUrlParams']);
+    //return base url with the appended extra url params;
+    return expandExtraUrlParams(ampdoc, params, expansionOption, bindings)
+        .then(extraUrlParams => {
+          return defaultBatchPlugin(expandedMsg, [{extraUrlParams}]);
+        });
   });
 }
 
 /**
  * Function that handler extraUrlParams from config and trigger.
- * @param {!./variables.VariableService} variableService
- * @param {?JsonObject} configParams
- * @param {?JsonObject} triggerParams
+ * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+ * @param {!Object} params
  * @param {!./variables.ExpansionOptions} expansionOption
- * @return {!Promise<!JsonObject>}
+ * @param {!Object} bindings
+ * @param {!Object=} opt_whitelist
+ * @return {!Promise<!Object>}
  * @private
  */
-function getExtraUrlParams(
-  variableService, configParams, triggerParams, expansionOption) {
+function expandExtraUrlParams(
+  ampdoc, params, expansionOption, bindings, opt_whitelist) {
+  const variableService = variableServiceFor(ampdoc.win);
+  const urlReplacements = Services.urlReplacementsForDoc(ampdoc);
+
   const requestPromises = [];
-  const params = map();
   // Don't encode param values here,
   // as we'll do it later in the getExtraUrlParamsString call.
   const option = new ExpansionOptions(
       expansionOption.vars,
       expansionOption.iterations,
       true /* noEncode */);
   // Add any given extraUrlParams as query string param
-  if (configParams || triggerParams) {
-    Object.assign(params, configParams, triggerParams);
-    for (const k in params) {
-      if (typeof params[k] == 'string') {
-        requestPromises.push(
-            variableService.expandTemplate(params[k], option)
-                .then(value => { params[k] = value; }));
-      }
-    }
-  }
-  return Promise.all(requestPromises).then(() => {
-    return params;
-  });
-}
-
-/**
- * Handle the params map and form the final extraUrlParams string
- * @param {!Object} params
- * @return {string}
- */
-function getExtraUrlParamsString(params) {
-  const s = [];
   for (const k in params) {
-    const v = params[k];
-    if (v == null) {
-      continue;
-    } else {
-      const sv = encodeVars(v);
```

Comment:
bad diff made it looks complicated... `encodeVars` is removed. we use `serializeQueryString` now.

I suggest you ignore the diff and take a look at the new code alone to get a sense what's happening. 

the main reason `encodeVars` was needed is because we need to keep macros unencoded during serialization. but because now we resolved all macro before serialization, we don;t need it anymore, hence a lot of code gets deleted.

Reply:
I looked at the diff and the new code and the old code for an hour, I believe I know what's happening here. 

What I see is that before we call `expandUrlAsync` with encoded string, while now we don't. And I want to make sure that won't change behavior.

Reply:
Talked offline. It may lead to a behavior change, but the new behavior is the expected one

Reply:
discussed offline, expand before encode is the right way to go